File: hivesim-rs/src/testapi.rs
use crate::types::{ClientDefinition, SuiteID, TestID, TestResult};
use crate::Simulation;
use ::std::{boxed::Box, future::Future, pin::Pin};
use async_trait::async_trait;
use core::fmt::Debug;
use dyn_clone::DynClone;
use jsonrpsee::http_client::{HttpClient, HttpClientBuilder};
use std::net::IpAddr;

use crate::utils::{client_test_name, extract_test_results};

pub type AsyncTestFunc = fn(
    &mut Test,
    Option<Client>,
) -> Pin<
    Box<
        dyn Future<Output = ()> // future API / pollable
            + Send // required by non-single-threaded executors
            + '_,
    >,
>;

pub type AsyncClientTestFunc = fn(
    Client,
) -> Pin<
    Box<
        dyn Future<Output = ()> // future API / pollable
            + Send // required by non-single-threaded executors
            + 'static,
    >,
>;

pub type AsyncTwoClientsTestFunc = fn(
    Client,
    Client,
) -> Pin<
    Box<
        dyn Future<Output = ()> // future API / pollable
            + Send // required by non-single-threaded executors
            + 'static,
    >,
>;

pub type AsyncNClientsTestFunc = fn(
    Vec<Client>,
) -> Pin<
    Box<
        dyn Future<Output = ()> // future API / pollable
            + Send // required by non-single-threaded executors
            + 'static,
    >,
>;

#[async_trait]
pub trait Testable: DynClone + Send + Sync {
    async fn run_test(&self, simulation: Simulation, suite_id: SuiteID, suite: Suite);
}

impl Debug for dyn Testable {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "Testable")
    }
}

dyn_clone::clone_trait_object!(Testable);
/// Description of a test suite
#[derive(Clone, Debug)]
pub struct Suite {
    pub name: String,
    pub description: String,
    pub tests: Vec<Box<dyn Testable>>,
}

impl Suite {
    pub fn add<T: Testable + 'static>(&mut self, test: T) {
        self.tests.push(Box::new(test))
    }
}

/// Represents a running client.
#[derive(Debug, Clone)]
pub struct Client {
    pub kind: String,
    pub container: String,
    pub ip: IpAddr,
    pub rpc: HttpClient,
    pub test: Test,
}

#[derive(Clone, Debug)]
pub struct TestRun {
    pub suite_id: SuiteID,
    pub suite: Suite,
    pub name: String,
    pub desc: String,
    pub always_run: bool,
}

/// A running test
#[derive(Clone, Debug)]
pub struct Test {
    pub sim: Simulation,
    pub test_id: TestID,
    pub suite: Suite,
    pub suite_id: SuiteID,
    pub result: TestResult,
}

impl Test {
    pub async fn start_client(&self, client_type: String) -> Client {
        let (container, ip) = self
            .sim
            .start_client(self.suite_id, self.test_id, client_type.clone())
            .await;

        let rpc_url = format!("http://{}:8545", ip);

        let rpc_client = HttpClientBuilder::default().build(rpc_url).unwrap();

        Client {
            kind: client_type,
            container,
            ip,
            rpc: rpc_client,
            test: Test {
                sim: self.sim.clone(),
                test_id: self.test_id,
                suite: self.suite.clone(),
                suite_id: self.suite_id,
                result: self.result.clone(),
            },
        }
    }

    /// Runs a subtest of this test.
    pub async fn run(&self, spec: impl Testable) {
        spec.run_test(self.sim.clone(), self.suite_id, self.suite.clone())
            .await
    }
}

/// ClientTestSpec is a test against a single client.
/// When used as a test in a suite, the test runs against all available client types.
#[derive(Clone)]
pub struct ClientTestSpec {
    // These fields are displayed in the UI. Be sure to add
    // a meaningful description here.
    pub name: String,
    // If AlwaysRun is true, the test will run even if Name does not match the test
    // pattern. This option is useful for tests that launch a client instance and
    // then perform further tests against it.
    pub description: String,
    // If AlwaysRun is true, the test will run even if Name does not match the test
    // pattern. This option is useful for tests that launch a client instance and
    // then perform further tests against it.
    pub always_run: bool,
    // The Run function is invoked when the test executes.
    pub run: AsyncClientTestFunc,
}

#[async_trait]
impl Testable for ClientTestSpec {
    async fn run_test(&self, simulation: Simulation, suite_id: SuiteID, suite: Suite) {
        let clients = simulation.client_types().await;

        for client in clients {
            let client_name = client.name;
            let test_run = TestRun {
                suite_id,
                suite: suite.clone(),
                name: client_test_name(self.name.clone(), client_name.clone()),
                desc: self.description.clone(),
                always_run: self.always_run,
            };

            run_client_test(simulation.clone(), test_run, client_name, self.run).await;
        }
    }
}

async fn run_client_test(
    host: Simulation,
    test: TestRun,
    client_name: String,
    func: AsyncClientTestFunc,
) {
    // Register test on simulation server and initialize the Test.
    let test_id = host.start_test(test.suite_id, test.name, test.desc).await;
    let suite_id = test.suite_id;

    let mut test = &mut Test {
        sim: host.clone(),
        test_id,
        suite: test.suite,
        suite_id,
        result: Default::default(),
    };

    test.result.pass = true;

    let client = test.start_client(client_name).await;

    // run test function
    let test_result = extract_test_results(
        tokio::spawn(async move {
            (func)(client).await;
        })
        .await,
    );

    host.end_test(suite_id, test_id, test_result).await;
}

#[derive(Clone)]
pub struct TestSpec {
    // These fields are displayed in the UI. Be sure to add
    // a meaningful description here.
    pub name: String,
    pub description: String,
    // If AlwaysRun is true, the test will run even if Name does not match the test
    // pattern. This option is useful for tests that launch a client instance and
    // then perform further tests against it.
    pub always_run: bool,
    // The Run function is invoked when the test executes.
    pub run: AsyncTestFunc,
    pub client: Option<Client>,
}

#[async_trait]
impl Testable for TestSpec {
    async fn run_test(&self, simulation: Simulation, suite_id: SuiteID, suite: Suite) {
        let test_run = TestRun {
            suite_id,
            suite,
            name: self.name.to_owned(),
            desc: self.description.to_owned(),
            always_run: self.always_run,
        };

        run_test(simulation, test_run, self.client.clone(), self.run).await;
    }
}

pub async fn run_test(
    host: Simulation,
    test: TestRun,
    client: Option<Client>,
    func: AsyncTestFunc,
) {
    // Register test on simulation server and initialize the T.
    let test_id = host.start_test(test.suite_id, test.name, test.desc).await;

    let mut test = &mut Test {
        sim: host.clone(),
        test_id,
        suite: test.suite,
        suite_id: test.suite_id,
        result: Default::default(),
    };

    test.result.pass = true;

    // run test function
    (func)(test, client).await;

    host.end_test(test.suite_id, test_id, test.result.clone())
        .await;
}

#[derive(Clone)]
pub struct TwoClientTestSpec<'a> {
    // These fields are displayed in the UI. Be sure to add
    // a meaningful description here.
    pub name: String,
    pub description: String,
    // If AlwaysRun is true, the test will run even if Name does not match the test
    // pattern. This option is useful for tests that launch a client instance and
    // then perform further tests against it.
    pub always_run: bool,
    // The Run function is invoked when the test executes.
    pub run: AsyncTwoClientsTestFunc,
    pub client_a: &'a ClientDefinition,
    pub client_b: &'a ClientDefinition,
}

#[async_trait]
impl Testable for TwoClientTestSpec<'_> {
    async fn run_test(&self, simulation: Simulation, suite_id: SuiteID, suite: Suite) {
        let test_run = TestRun {
            suite_id,
            suite,
            name: self.name.to_owned(),
            desc: self.description.to_owned(),
            always_run: self.always_run,
        };

        run_two_client_test(simulation, test_run, self.client_a, self.client_b, self.run).await;
    }
}

// Write a test that runs against two clients.
async fn run_two_client_test(
    host: Simulation,
    test: TestRun,
    client_a: &ClientDefinition,
    client_b: &ClientDefinition,
    func: AsyncTwoClientsTestFunc,
) {
    // Register test on simulation server and initialize the T.
    let test_id = host.start_test(test.suite_id, test.name, test.desc).await;
    let suite_id = test.suite_id;

    let mut test = &mut Test {
        sim: host.clone(),
        test_id,
        suite: test.suite,
        suite_id,
        result: Default::default(),
    };

    test.result.pass = true;

    let client_a = test.start_client(client_a.name.clone()).await;
    let client_b = test.start_client(client_b.name.clone()).await;

    // run test function
    let test_result = extract_test_results(
        tokio::spawn(async move {
            (func)(client_a, client_b).await;
        })
        .await,
    );

    host.end_test(suite_id, test_id, test_result).await;
}

#[derive(Clone)]
pub struct NClientTestSpec<'a> {
    // These fields are displayed in the UI. Be sure to add
    // a meaningful description here.
    pub name: String,
    pub description: String,
    // If AlwaysRun is true, the test will run even if Name does not match the test
    // pattern. This option is useful for tests that launch a client instance and
    // then perform further tests against it.
    pub always_run: bool,
    // The Run function is invoked when the test executes.
    pub run: AsyncNClientsTestFunc,
    pub clients: &'a Vec<&'a ClientDefinition>,
}

#[async_trait]
impl Testable for NClientTestSpec<'_> {
    async fn run_test(&self, simulation: Simulation, suite_id: SuiteID, suite: Suite) {
        let test_run = TestRun {
            suite_id,
            suite,
            name: self.name.to_owned(),
            desc: self.description.to_owned(),
            always_run: self.always_run,
        };

        run_n_client_test(simulation, test_run, self.clients, self.run).await;
    }
}

// Write a test that runs against N clients.
async fn run_n_client_test(
    host: Simulation,
    test: TestRun,
    clients: &Vec<&ClientDefinition>,
    func: AsyncNClientsTestFunc,
) {
    // Register test on simulation server and initialize the T.
    let test_id = host.start_test(test.suite_id, test.name, test.desc).await;
    let suite_id = test.suite_id;

    let mut test = &mut Test {
        sim: host.clone(),
        test_id,
        suite: test.suite,
        suite_id,
        result: Default::default(),
    };

    test.result.pass = true;

    let mut client_vec: Vec<Client> = Vec::new();
    for i in clients {
        client_vec.push(test.start_client(i.name.clone()).await);
    }

    // run test function
    let test_result = extract_test_results(
        tokio::spawn(async move {
            (func)(client_vec).await;
        })
        .await,
    );

    host.end_test(suite_id, test_id, test_result).await;
}

File: hivesim-ts/testapi.ts
import jayson from "jayson/promise";
import { Simulation } from "./simulation.js";
import { ClientDefinition, SuiteID, TestID, TestResult } from "./types";
import { client_test_name } from "./utils.js";

const { Client } = jayson;
type HttpClient = jayson.HttpClient;
type AsyncTestFunc = (test: Test, client?: IClient) => Promise<void>;
type AsyncClientTestFunc = (test: Test, client: IClient) => Promise<void>;
type AsyncTwoTestFunc = (
  test: Test,
  client_a?: IClient,
  client_b?: IClient
) => Promise<void>;
type AsyncTwoClientsTestFunc = (
  test: Test,
  client_a: IClient,
  client_b: IClient
) => Promise<void>;
type AsyncNClientsTestFunc = (test: Test, clients: IClient[]) => Promise<void>;

type TestFunc =
  | AsyncTestFunc
  | AsyncClientTestFunc
  | AsyncTwoClientsTestFunc
  | AsyncNClientsTestFunc;
export interface Testable {
  name: string;
  description: string;
  always_run: boolean;
  run: TestFunc;
  run_test: (
    simulation: Simulation,
    suite_id: SuiteID,
    suite: Suite
  ) => Promise<any>;
}


export class Suite {
  name: string;
  description: string;
  tests: Testable[];
  constructor(name: string, description: string) {
    this.name = name;
    this.description = description;
    this.tests = [];
  }
  add(test: Testable) {
    this.tests.push(test);
  }
  // // RunSuite runs all tests in a suite.
  async run(host: Simulation) {
    const suiteId = await host.start_suite(this.name, this.description);
    for await (const test of this.tests) {
      await test.run_test(host, suiteId, this);
    }
    await host.end_suite(suiteId);
  }

  async mustRunSuite(host: Simulation): Promise<void> {
    try {
      await this.run(host);
    } catch (err: any) {
      throw err;
    }
  }
}

export interface IHttpClient {
  transport: HttpClient;
  request_timeout: number;
}
export interface IClient {
  kind: string;
  container: string;
  ip: string;
  rpc: HttpClient;
  test: ITest;
}
export interface ITestRun {
  suite_id: SuiteID;
  suite: Suite;
  name: string;
  desc: string;
  always_run: boolean;
}

export interface ITest {
  sim: Simulation;
  suite_id: SuiteID;
  suite: Suite;
  test_id: TestID;
  result: TestResult;
}

export class Test {
  sim: Simulation;
  suite: Suite;
  suite_id: SuiteID;
  test_id: TestID;
  result: TestResult;

  constructor(
    sim: Simulation,
    suite_id: SuiteID,
    suite: Suite,
    test_id: TestID,
    result: TestResult = {
      details: "",
      pass: false,
    }
  ) {
    this.sim = sim;
    this.suite_id = suite_id;
    this.suite = suite;
    this.test_id = test_id;
    this.result = result;
  }

  async start_client(client_type: string): Promise<IClient> {
    const [container, ip] = await this.sim.start_client(
      this.suite_id,
      this.test_id,
      client_type
    );
    const rpc_client = Client.http({ port: 8545, host: ip });
    const client: IClient = {
      kind: client_type,
      container: container,
      ip: ip,
      rpc: rpc_client,
      test: {
        sim: this.sim,
        suite_id: this.suite_id,
        suite: this.suite,
        test_id: this.test_id,
        result: this.result!,
      },
    };
    return client;
  }

  async run(spec: Testable) {
    await spec.run_test(this.sim, this.suite_id, this.suite);
  }

  fatal(message: string) {
    this.log_failure(message);
    this.fail();
  }

  log_failure(message: string) {
    this.result.details = message;
  }

  fail() {
    this.result.pass = false;
  }
}
interface ITestSpec extends Testable {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncTestFunc;
  client?: IClient;
}

interface IClientTestSpec extends Testable {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncClientTestFunc;
}
interface I2ClientTestSpec extends Testable {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncTwoClientsTestFunc;
}
export class ClientTestSpec implements IClientTestSpec {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncClientTestFunc;
  constructor(opts: {
    name: string;
    description: string;
    always_run: boolean;
    run: AsyncClientTestFunc;
  }) {
    this.name = opts.name;
    this.description = opts.description;
    this.always_run = opts.always_run;
    this.run = opts.run;
  }
  async run_test(sim: Simulation, suite_id: SuiteID, suite: Suite) {
    const clients = await sim.client_types();
    for (const client of clients) {
      const client_name = client.name;
      const name = client_test_name(this.name, client_name);
      const test_run: ITestRun = {
        suite_id,
        suite,
        name,
        desc: this.description,
        always_run: this.always_run,
      };
      await this.run_client_test(sim, test_run, client_name, this.run);
    }
  }
  async run_client_test(
    host: Simulation,
    test_run: ITestRun,
    client_name: string,
    run: AsyncClientTestFunc
  ) {
    const test_id = await host.start_test(
      test_run.suite_id,
      test_run.name,
      test_run.desc
    );
    const test: Test = new Test(
      host,
      test_run.suite_id,
      test_run.suite,
      test_id
    );
    test.result.pass = true;

    const client = await test.start_client(client_name);
    await run(test, client);
    const end = await host.end_test(test);
  }
}

export class TestSpec implements ITestSpec {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncTestFunc;
  client?: IClient;
  constructor(opts: {
    name: string;
    description: string;
    always_run: boolean;
    run: AsyncTestFunc;
    client?: IClient;
  }) {
    this.name = opts.name;
    this.description = opts.description;
    this.always_run = opts.always_run;
    this.run = opts.run;
    this.client = opts.client;
  }

  async run_test(simulation: Simulation, suite_id: SuiteID, suite: Suite) {
    const test_run: ITestRun = {
      suite_id: suite_id,
      suite: suite,
      name: this.name,
      desc: this.description,
      always_run: this.always_run,
    };
    await run_test(simulation, test_run, this.client, this.run);
  }
}

export const run_test = async (
  host: Simulation,
  test: ITestRun,
  client: IClient | undefined,
  func: AsyncTestFunc
) => {
  const test_id = await host.start_test(test.suite_id, test.name, test.desc);
  const t: Test = new Test(host, test.suite_id, test.suite, test_id);
  t.result.pass = true;

  await func(t, client);

  await host.end_test(t);
};

export class TwoClientTestSpec implements I2ClientTestSpec {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncTwoClientsTestFunc;
  constructor(opts: {
    name: string;
    description: string;
    always_run: boolean;
    run: AsyncTwoClientsTestFunc;
  }) {
    this.name = opts.name;
    this.description = opts.description;
    this.always_run = opts.always_run;
    this.run = opts.run;
  }

  async run_test(sim: Simulation, suite_id: SuiteID, suite: Suite) {
    const clients = await sim.client_types();
    for (const client of clients) {
      const client_name = client.name;
      for (const client2 of clients) {
        const client_2_name = client2.name;
        const name = client_test_name(this.name, client_name);
        const test_run: ITestRun = {
          suite_id,
          suite,
          name: name + " --> " + client_2_name,
          desc: this.description,
          always_run: this.always_run,
        };
        await this.run_2_client_test(
          sim,
          test_run,
          client_name,
          client_2_name,
          this.run
        );
      }
    }
  }


  async run_2_client_test(
    host: Simulation,
    test_run: ITestRun,
    client_a_name: string,
    client_b_name: string,
    run: AsyncTwoClientsTestFunc
  ) {
    const test_id = await host.start_test(
      test_run.suite_id,
      test_run.name,
      test_run.desc
    );
    const test: Test = new Test(
      host,
      test_run.suite_id,
      test_run.suite,
      test_id
    );
    test.result.pass = true;

    const client_a = await test.start_client(client_a_name);
    const client_b = await test.start_client(client_b_name);
    await run(test, client_a, client_b);
    const end = await host.end_test(test);
  }
}



export class NClientTestSpec implements Testable {
  name: string;
  description: string;
  always_run: boolean;
  run: AsyncNClientsTestFunc;
  clients: ClientDefinition[];
  constructor(
    name: string,
    description: string,
    always_run: boolean,
    run: AsyncNClientsTestFunc,
    clients: ClientDefinition[]
  ) {
    this.name = name;
    this.description = description;
    this.always_run = always_run;
    this.run = run;
    this.clients = clients;
  }
  async run_test(simulation: Simulation, suite_id: SuiteID, suite: Suite) {
    const test_run: ITestRun = {
      suite_id: suite_id,
      suite: suite,
      name: this.name,
      desc: this.description,
      always_run: this.always_run,
    };
    await run_n_client_test(simulation, test_run, this.clients, this.run);
  }
}

export const run_n_client_test = async (
  host: Simulation,
  test_run: ITestRun,
  clients: ClientDefinition[],
  func: AsyncNClientsTestFunc
) => {
  const test_id = await host.start_test(
    test_run.suite_id,
    test_run.name,
    test_run.desc
  );
  const test: Test = new Test(host, test_run.suite_id, test_run.suite, test_id);
  test.result.pass = true;
  const client_vec: IClient[] = await Promise.all(
    clients.map((client) => {
      return test.start_client(client.name);
    })
  );
  for (const client of clients) {
    await test.start_client(client.name);
  }
  await func(test, client_vec);
  await host.end_test(test);
};
